crates/core/src/store.rs

```diff
@@ -410,6 +432,19 @@ impl StoreBuilder {
     }
 }
 
+fn with_preview2_ctx(
+    ctx: &mut wasi_preview2::WasiCtxBuilder,
+    dummy: wasi_preview2::WasiCtxBuilder,
+    f: impl FnOnce(wasi_preview2::WasiCtxBuilder) -> wasi_preview2::WasiCtxBuilder,
+) -> wasi_preview2::WasiCtxBuilder {
+    let ctx_addr = std::ptr::addr_of!(*ctx);
+    let owned = std::mem::replace(ctx, dummy);
+    let owned = f(owned);
+    let dummy = std::mem::replace(ctx, owned);
+    assert_eq!(ctx_addr, std::ptr::addr_of!(*ctx));
+    dummy
+}
+
 /// A builder of a `WasiCtx` for all versions of Wasi
 #[allow(clippy::large_enum_variant)]
 enum WasiCtxBuilder {
```

Comment:
The conclusion of https://github.com/bytecodealliance/wasmtime/pull/6770 was that this should probably use a `WasiCtxBuilder` in both arms to use the move-style everywhere instead of on just one half to avoid some of the weirdness here. Otherwise the other possible fix is to update to `&mut`-style builders but that has other API drawbacks that aren't a clear win at this time.

The major difference from preview1 is that the preview2 `WasiCtx` doesn't have the mutators that preview1's context has I believe.

Reply:
I commented over there, because I'm still not entirely convinced, but we can certainly live with things as they are now

Reply:
I'm also not too sure how using a builder in both arms would actually make things easier, but I'm probably missing something there